cpp/ql/src/experimental/Security/CWE/CWE-193/ConstantSizeArrayOffByOne.ql

```diff
@@ -93,56 +78,61 @@ predicate isInvalidPointerDerefSink2(DataFlow::Node sink, Instruction i, string
   )
 }
 
-predicate isConstantSizeOverflowSource(Field f, FieldAddressToPointerArithmeticFlow::PathNode fieldSource, PointerAddInstruction pai, int delta) {
-  exists(int size, int bound, FieldAddressToPointerArithmeticFlow::PathNode sink |
-    FieldAddressToPointerArithmeticFlow::flowPath(fieldSource, sink) and
-    isFieldAddressSource(f, fieldSource.getNode()) and
-    pai.getLeft() = sink.getNode().(DataFlow::InstructionNode).asInstruction() and
-    pai.getElementSize() = f.getUnspecifiedType().(ArrayType).getBaseType().getSize() and
-    f.getUnspecifiedType().(ArrayType).getArraySize() = size and
-    semBounded(getSemanticExpr(pai.getRight()), any(SemZeroBound b), bound, true, _) and
-    delta = bound - size and
-    delta >= 0 and
-    size != 0 and
-    size != 1
-  )
+predicate pointerArithOverflow(
+  PointerArithmeticInstruction pai, Field f, int size, int bound, int delta
+) {
+  pai.getElementSize() = f.getUnspecifiedType().(ArrayType).getBaseType().getSize() and
+  f.getUnspecifiedType().(ArrayType).getArraySize() = size and
+  semBounded(getSemanticExpr(pai.getRight()), any(SemZeroBound b), bound, true, _) and
+  delta = bound - size
 }
 
-module PointerArithmeticToDerefConfig implements DataFlow::ConfigSig {
-  predicate isSource(DataFlow::Node source) {
-    isConstantSizeOverflowSource(_, _, source.asInstruction(), _)
-  }
+module FieldAddressToDerefConfig implements DataFlow::StateConfigSig {
+  newtype FlowState =
+    additional TArray(Field f) or
+    additional TOverflowArithmetic(PointerArithmeticInstruction pai)
```

Comment:
We might as well restrict `f` to be a field that satisfies `pointerArithOverflow(_, f, _, _, _)`, right?

And similarly: We might as well restrict `pai` to be an instruction that satisfies `pointerArithOverflow(pai, _, _, _, _)`, right?